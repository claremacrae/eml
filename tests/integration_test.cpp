#include <ApprovalTests.hpp>
#include <catch2/catch.hpp>

#include "eml.hpp"

void verify_eml(const std::string& source)
{
  eml::GarbageCollector gc{};

  eml::CompilerConfig config = {eml::SameScopeShadowing::allow};
  eml::Compiler compiler{gc, config};
  eml::VM vm{gc};

  std::stringstream ss;
  ss << "Input:\n" << source << "\n\n";

  if (!source.empty()) {
    compiler.compile(source)
        .map([&vm, &ss](auto tuple) {
          const auto [bytecode, type] = tuple;

          ss << "Compile to:\n" << bytecode << '\n';

          const auto result = vm.interpret(bytecode);
          REQUIRE(result);
          ss << "Executes to:\n" << eml::to_string(type, *result) << '\n';
        })
        .map_error([&ss](const auto& errors) {
          ss << "Error message:\n";
          std::for_each(std::begin(errors), std::end(errors),
                        [&ss](auto e) { ss << eml::to_string(e); });
        });
    ApprovalTests::Approvals::verify(ss.str());
  }
}

void verify_for_documentation(const std::string& source)
{
  eml::GarbageCollector gc{};

  eml::CompilerConfig config = {eml::SameScopeShadowing::allow};
  eml::Compiler compiler{gc, config};
  eml::VM vm{gc};

  std::stringstream ss;
  ss << "Input:\n" << source << "\n\n";

  if (!source.empty()) {
    compiler.compile(source)
        .map([&vm, &ss](auto tuple) {
          const auto [bytecode, type] = tuple;

          const auto result = vm.interpret(bytecode);
          REQUIRE(result);
          ss << "Executes to:\n" << eml::to_string(type, *result) << '\n';
        })
        .map_error([&ss](const auto& errors) {
          ss << "Error message:\n";
          std::for_each(std::begin(errors), std::end(errors),
                        [&ss](auto e) { ss << eml::to_string(e); });
        });
    ApprovalTests::Approvals::verify(
        ss.str(),
        ApprovalTests::Options{}.fileOptions().withFileExtension(".docs.txt"));
  }
}

TEST_CASE("Integration Test")
{
  // begin-snippet: example_from_code
  const std::string s = "if (5 > 1) {2 + 3} else {4 - 6}";
  // end-snippet
  verify_eml(s);
  verify_for_documentation(s);
}

TEST_CASE("Test Error")
{
  verify_eml("if (5 > 1 {2 + 3} else {4 - 6}");
}

TEST_CASE("Test with String")
{
  const std::string s = R"("Hello, " ++ "world")";
  verify_eml(s);
}

TEST_CASE("Test with Arithmetics")
{
  verify_eml("1 + 3 / 4 * 5 - 10");
}