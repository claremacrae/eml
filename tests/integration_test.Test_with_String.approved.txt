Input:
"Hello, " ++ "world"

Compile to:
0000    01 00          push 0 //4.64134e-310
   |    01 01          push 1 //4.64134e-310
   |    0c             string_cat

Executes to:
"Hello, world": String

